### ISSUE TITLE
Strange behaviour with sub() / gsub() and Unicode characters

### ISSUE THREAD

**Issue by okdana** [jqlang/jq#1166]
Hello! I've been trying to use `gsub()` to collapse consecutive white-space in article titles in my input, and that worked fine... until i came across a title containing a 'smart' apostrophe, and i found that my replacement wasn't working as expected:

```
% jq '.foo | sub("\\s+"; "_")' <<< '{ "foo": "bar\u2019s baz" }'
"_ baz" # Expected: "bar’s_baz"
```

Basically any time a 'smart' possessive appears in one of these titles (_Father’s Day_ for example), the entire word containing the possessive is destroyed.

Following some experimentation i discovered that the problem only seems to occur with 3+ byte sequences (`U+0800` and above):

```
% jq '.foo | sub("\\s+"; "_")' <<< '{ "foo": "bar\u07ffs baz" }'
"bar߿s_baz"

% jq '.foo | sub("\\s+"; "_")' <<< '{ "foo": "bar\u0800s baz" }'
"_ baz"
```

I'm not misunderstanding something super obvious here, am i?

**Comment by wtlangford** [comment]
Smart quotation marks are the bane of my existence.

I'm not around for most of today, but can you try your test case on master?
I think it might be a byte-length vs character-length issue. Or it's
something weird in the library.

**Comment by okdana** [comment]
Oh, i'm sorry, i should have done that before i posted.

It is totally fixed already in `master`:

```
% ./jq -V
jq-1.5rc2-152-g0b82185

% ./jq '.foo | sub("\\s+"; "_")' <<< '{ "foo": "bar\u2019s baz" }'
"bar’s_baz"
```

I guess that's resolved for me! Do you know if there's any estimate as to when an official build with that fix might see release?

**Comment by danilobellini** [comment]
Same here:

```shell
$ jq '.foo | sub("\\s+"; "_")' <<< '{ "foo": "bar\u2019s baz" }'
"_ baz"
```

It's the Arch Linux version:

```shell
$ jq -V
jq-1.5
$ pacman -Q jq
jq 1.5-5
```

As it seems, there's something buggy in the index calculation:

```shell
$ jq -a '.foo | indices("b")' <<< '{ "foo": "bar\u2019s baz" }'
[
  0,
  8
]
$ jq -a '.foo | rindex("b")' <<< '{ "foo": "bar\u2019s baz" }'
8
```

Even creating myself a substring replacement function gave me trouble due to that shifted index.

**Comment by pkoppstein** [comment]
@danilobellini - The problem was fixed (long ago) but you will need to get the "master" version of jq to see the fix:

    $ jq -V 
    jq-1.5rc2-150-g1740fd0
    $ jq '.foo | sub("\\s+"; "_")' <<< '{ "foo": "bar\u2019s baz" }'
    "bar’s_baz"

**Comment by danilobellini** [comment]
Is that commit version more stable than the 1.5 release? Why it isn't tagged/released as 1.5.1 yet? I hadn't tried that version because its version string made me believe it was a release candidate for the version 1.5, not something new.

Perhaps I should flag [its Arch Linux package](https://www.archlinux.org/packages/community/x86_64/jq/) as out of date instead of just installing it myself here. It [already had a patch](https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/jq&id=4455ecb03f020b9273c319714c34fd18393e4791), so I was wondering which commits should be patched / "cherry-picked" to fix this, then I got this:

```
$ git clone https://github.com/stedolan/jq
Cloning into 'jq'...
remote: Counting objects: 6595, done.
remote: Total 6595 (delta 0), reused 0 (delta 0), pack-reused 6595
Receiving objects: 100% (6595/6595), 4.93 MiB | 259.00 KiB/s, done.
Resolving deltas: 100% (3924/3924), done.

$ cd jq

$ git checkout -b distro jq-1.5
Switched to a new branch 'distro'

$ git cherry-pick 8eb1367ca44e772963e704a700ef72ae2e12babd
[distro a140d51] Heap buffer overflow in tokenadd() (fix #105)
 Author: Nicolas Williams <nico@cryptonector.com>
 Date: Sat Oct 24 17:24:57 2015 -0500
 1 file changed, 2 insertions(+), 2 deletions(-)

$ git cherry-pick 6c3934de158d54672f94a1bfc1227b761d4caa10
[distro 7c94afe] Fix decoding of UTF-8 sequence length (fix #922)
 Author: David Tolnay <dtolnay@gmail.com>
 Date: Sat Aug 22 12:18:13 2015 -0700
 2 files changed, 10 insertions(+), 4 deletions(-)

$ autoreconf -f # Required as I have aclocal-1.15, not aclocal-1.14

$ ./configure --prefix=/usr
[...]

$ make
[...]

$ ./jq '.foo | sub("\\s+"; "_")' <<< '{ "foo": "bar\u2019s baz" }'
"bar’s_baz"
```

=)

However...

```
$ ./jq -a '.foo | indices("b")' <<< '{ "foo": "bar\u2019s baz" }'
[
  0,
  8
]

$ ./jq -a '.foo | rindex("b")' <<< '{ "foo": "bar\u2019s baz" }'
8
```

... the indices are still 8, not 6. The same happens on the master branch (commit 0b8218515eabf1a967eba0dbcc7a0e5ae031a509) made with `autoreconf --install && ./configure --prefix=/usr && make`.

**Comment by pkoppstein** [comment]
1. The current tip of master (0b82185) has not changed since March 21, 2016, and in this and other senses should be considered stable.  I'd certainly recommend that everyone who can gain access to it should use it, but it's true (and unfortunate) that it isn't an official "release". 

2. The "index" family of filters (index, indices, rindex) do not have the semantics you evidently expect.  To find the "codepoint-index", I'd use `match/1` or `match/2`. Consider:

```
$ jq -V
jq-1.5rc2-150-g1740fd0

$ jq '.foo | match("b";"g")' <<< '{ "foo": "bar\u2019s baz" }'

{
  "offset": 0,
  "length": 1,
  "string": "b",
  "captures": []
}
{
  "offset": 6,
  "length": 1,
  "string": "b",
  "captures": []
}


$ jq '.foo | indices("b")' <<< '{ "foo": "bar\u2019s baz" }'

[
  0,
  8
]
```

**Comment by moll** [comment]
Also stumbled on `sub` messing up wrong characters a few times when a multibyte character is involved. Any luck getting this to master and eventually to distros?

**Comment by itchyny** [comment]
This bug was already fixed and released in [1.6](https://github.com/jqlang/jq/releases/tag/jq-1.6) in Nov 2018, so closing this issue.
